projects/packages/identity-crisis/src/class-rest-endpoints.php

```diff
@@ -0,0 +1,164 @@
+				'permission_callback' => __CLASS__ . '::identity_crisis_mitigation_permission_check',
+			)
+		);
+
+		// IDC resolve: create an entirely new shadow site for this URL.
+		register_rest_route(
+			'jetpack/v4',
+			'/identity-crisis/start-fresh',
+			array(
+				'methods'             => WP_REST_Server::EDITABLE,
+				'callback'            => __CLASS__ . '::start_fresh_connection',
+				'permission_callback' => __CLASS__ . '::identity_crisis_mitigation_permission_check',
+			)
+		);
+
+	}
+
+	/**
+	 * Handles identity crisis mitigation, confirming safe mode for this site.
+	 *
+	 * @since 4.4.0
+	 *
+	 * @return bool | WP_Error True if option is properly set.
+	 */
+	public static function confirm_safe_mode() {
+		$updated = Jetpack_Options::update_option( 'safe_mode_confirmed', true );
+		if ( $updated ) {
+			return rest_ensure_response(
+				array(
+					'code' => 'success',
+				)
+			);
+		}
+
+		return new WP_Error(
+			'error_setting_jetpack_safe_mode',
+			esc_html__( 'Could not confirm safe mode.', 'jetpack' ),
+			array( 'status' => 500 )
+		);
+	}
+
+	/**
+	 * Handles identity crisis mitigation, migrating stats and subscribers from old url to this, new url.
+	 *
+	 * @since 4.4.0
+	 *
+	 * @return bool | WP_Error True if option is properly set.
+	 */
+	public static function migrate_stats_and_subscribers() {
+		if ( Jetpack_Options::get_option( 'sync_error_idc' ) && ! Jetpack_Options::delete_option( 'sync_error_idc' ) ) {
+			return new WP_Error(
+				'error_deleting_sync_error_idc',
+				esc_html__( 'Could not delete sync error option.', 'jetpack' ),
+				array( 'status' => 500 )
+			);
+		}
+
+		if ( Jetpack_Options::get_option( 'migrate_for_idc' ) || Jetpack_Options::update_option( 'migrate_for_idc', true ) ) {
+			return rest_ensure_response(
+				array(
+					'code' => 'success',
+				)
+			);
+		}
+		return new WP_Error(
+			'error_setting_jetpack_migrate',
+			esc_html__( 'Could not confirm migration.', 'jetpack' ),
+			array( 'status' => 500 )
+		);
+	}
+
+	/**
+	 * This IDC resolution will disconnect the site and re-connect to a completely new
+	 * and separate shadow site than the original.
+	 *
+	 * It will first will disconnect the site without phoning home as to not disturb the production site.
+	 * It then builds a fresh connection URL and sends it back along with the response.
+	 *
+	 * @since 4.4.0
+	 * @return bool|WP_Error
+	 */
+	public static function start_fresh_connection() {
+		/**
+		 * Fires when Users have requested through Identity Crisis for the connection to be reset.
+		 * Should be used to disconnect any connections and reset options.
+		 *
+		 * @since 9.9.0
+		 */
+		do_action( 'jetpack_idc_disconnect' );
+
+		/**
+		 * Filters the connection url that users should be redirected to for re-establishing their connection.
+		 *
+		 * @since 9.9.0
+		 *
+		 * @param string    $connection_url Connection URL user should be redirected to.
```

Comment:
I'm a little confused by this. The `Jetpack_Core_Json_Api_Endpoints::build_connect_url` docblock states that it returns a string or a `WP_Error` object, but that isn't correct. It returns a `WP_REST_Response` or a `WP_Error` object. So, this filter should probably also return the same.

Reply:
updated docblock and the filter name as it is different behavior now.